### PR TITLE
[types] add type annotations to osh/word_eval.py

### PR DESCRIPTION
**Pull request by asokoloski**
This went ok.  I'm getting the hang of it, I think.  I used "Any" in a few places that seemed appropriate.  Also had to add some `assert isinstance(<var>, <type>)` statements to make mypy happy.

### CODE REVIEW

File: osh/word_eval.py [hunk 1]
```diff
@@ -1488,7 +1666,7 @@ def EvalWordSequence2(self, words, allow_assign=False):
       argv: list of string arguments, or None if there was an eval error
     """
     if self.exec_opts.simple_word_eval:
-      return self.StaticEvalWordSequence2(words, allow_assign)
+      return self.StaticEvalWordSequence2(words, allow_assign) # follow
```
Comment (andychu) [inline osh/word_eval.py]: What is this comment for?

Also I use 2 spaces before #, although I'm won't  block on anything on that until we have an autoformatter...

Comment (asokoloski) [reply osh/word_eval.py]: Whoops!  That was just me trying to keep track while manually tracing through some calls.  I removed it (actually in 2 places): 37dcecc8c886c3ee8c0841d0cea86348e46dd5ff

Comment (asokoloski) [reply osh/word_eval.py]: For any future work I can jump right to using tagswitch.  Do you prefer that?

**Comment by asokoloski** [comment]
For any future work I can jump right to using tagswitch.  Should I do that?

**Comment by andychu** [comment]
Yeah if you think it's less total work then `with tagswitch()` is better.  It's already being used and translated in the parser so I assume it will work.

There's also this issue where I use `tag = node.tag_()`  in some places instead of `node.tag`.  That is also because of translation.  In C++  `tag_()` is an overloaded function, since you can have an "overloaded field".

I think I started off doing it in 2 steps because the first step is more automatable than the second (with some `sed` scripts).  But if neither step is automatable then it may be less work to combine them into one.  But use your discretion.

Thanks again!

**Comment by andychu** [comment]
To clarify, `tag_()` is used for some if statements.  `with tagswitch()` is translated directly to a switch statement, and does the check internally so you don't need `tag_()`.

**Comment by asokoloski** [comment]
Ok, thanks for clarifying!  And as always, glad to help :)

**Pull request merged**